src/multirust-cli/self_update.rs

```diff
@@ -0,0 +1,1000 @@
+
+    // Now, run a *system* binary to inherit the DELETE_ON_CLOSE
+    // handle to *this* process, then exit. The OS will delete the gc
+    // exe when it exits.
+    let rm_gc_exe = OsStr::new("net");
+    let mut rm_gc_exe_win: Vec<_> = rm_gc_exe.encode_wide().collect();
+    rm_gc_exe_win.push(0);
+
+    unsafe {
+        let mut si = mem::zeroed::<STARTUPINFOW>();
+        si.cb = mem::size_of::<STARTUPINFOW>() as DWORD;
+        let mut pi = mem::zeroed::<PROCESS_INFORMATION>();
+
+        // Launch the new process without stdio so it can't print to the console
+        si.dwFlags = STARTF_USESTDHANDLES;
+        si.hStdInput = ptr::null_mut();
+        si.hStdOutput = ptr::null_mut();
+        si.hStdError = ptr::null_mut();
+
+        let r = CreateProcessW(ptr::null(),
+                               rm_gc_exe_win.as_mut_ptr(),
+                               ptr::null_mut(), ptr::null_mut(),
+                               1, // inherit handles
+                               0,
+                               ptr::null_mut(), ptr::null(),
+                               &mut si, &mut pi);
+
+        if r == 0 {
+            let err = io::Error::last_os_error();
+            return Err(Error::WindowsUninstallMadness(err));
+        }
+    }
+
+    process::exit(0);
+}
+
+#[cfg(windows)]
+fn wait_for_parent() -> Result<()> {
+    use kernel32::{Process32First, Process32Next,
+                   CreateToolhelp32Snapshot, CloseHandle, OpenProcess,
+                   GetCurrentProcessId, WaitForSingleObject};
+    use winapi::{PROCESSENTRY32, INVALID_HANDLE_VALUE, DWORD,
+                 TH32CS_SNAPPROCESS, SYNCHRONIZE, WAIT_OBJECT_0};
+    use std::io;
+    use std::mem;
+
+    unsafe {
+        // Take a snapshot of system processes, one of which is ours
+        // and contains our parent's pid
+        let snapshot = CreateToolhelp32Snapshot(TH32CS_SNAPPROCESS, 0);
+        if snapshot == INVALID_HANDLE_VALUE {
+            let err = io::Error::last_os_error();
+            return Err(Error::WindowsUninstallMadness(err));
+        }
+
+        defer! {{ let _ = CloseHandle(snapshot); }}
+
+        let mut entry: PROCESSENTRY32 = mem::zeroed();
+        entry.dwSize = mem::size_of::<PROCESSENTRY32>() as DWORD;
+
+        // Iterate over system processes looking for ours
+        let success = Process32First(snapshot, &mut entry);
+        if success == 0 {
+            let err = io::Error::last_os_error();
+            return Err(Error::WindowsUninstallMadness(err));
+        }
+
+        let this_pid = GetCurrentProcessId();
+        while entry.th32ProcessID != this_pid {
+            let success = Process32Next(snapshot, &mut entry);
+            if success == 0 {
+                let err = io::Error::last_os_error();
+                return Err(Error::WindowsUninstallMadness(err));
+            }
+        }
+
+        let parent_id = entry.th32ParentProcessID;
```

Comment:
We discussed this, and while yeah, it might be, I don't want to rewrite it. The worse part is the process ID race, which we also agreed not to fix.